ttnn/cpp/ttnn/operations/ccl/all_gather/device/multi_core/all_gather_op_multi_core.cpp

```diff
@@ -29,6 +29,41 @@ namespace ttnn {
 
 using namespace ccl;
 
+static std::tuple<CoreRangeSet,CoreRangeSet> get_all_worker_cores(AllGatherConfig const& all_gather_config, uint32_t num_links, uint32_t num_full_send_directions, CoreCoord const& core_grid_offset) {
+    constexpr uint32_t worker_grid_width = 8;
+    const bool fit_sender_and_receiver_workers_on_same_row = (worker_grid_width / 2) >= all_gather_config.get_num_workers_per_link();
+    std::set<CoreRange> receiver_worker_cores = {};
+    std::set<CoreRange> sender_worker_cores = {};
+
+    for (uint32_t full_send_direction = 0; full_send_direction < num_full_send_directions; full_send_direction++) {
+        for (uint32_t link = 0; link < num_links; ++link) {
+            uint32_t max_cols = 8;
+            uint32_t curr_row = link * (((all_gather_config.get_num_workers_per_link()  * 2 - 1) / max_cols) + 1) +
+                (full_send_direction * num_links * (((all_gather_config.get_num_workers_per_link() * 2 - 1) / max_cols) + 1)) + core_grid_offset.y;
+            uint32_t curr_col = core_grid_offset.x;
+            for (uint32_t r = 0; r < all_gather_config.get_num_workers_per_link(); r++) {
+                receiver_worker_cores.insert(CoreRange(CoreCoord(curr_col, curr_row)));
+                curr_col ++;
+                if (curr_col == max_cols) {
+                    curr_col = 0;
+                    curr_row++;
+                }
+            }
+            for (uint32_t s = 0; s < all_gather_config.get_num_workers_per_link(); s++) {
+                sender_worker_cores.insert(CoreRange(CoreCoord(curr_col, curr_row)));
+                curr_col ++;
+                if (curr_col == max_cols) {
+                    curr_col = 0;
+                    curr_row++;
+                }
+            }
+        }
+    }
+
+    return {CoreRangeSet(receiver_worker_cores), CoreRangeSet(sender_worker_cores)};
+}
+
+
 static std::tuple<CoreRangeSet,CoreRangeSet> select_worker_cores(AllGatherConfig const& all_gather_config, uint32_t num_links, uint32_t link, uint32_t full_send_direction, CoreCoord const& core_grid_offset) {
```

Comment:
Can we delete this? I see it's used below but instead of 2 implementations, we should have just one to avoid bugs in the future. 

Some options:
- `get_all_worker_cores` returns an additional piece of information which is link/worker_index to core location
- both functions call same underlying code

(Personally I prefer the first option since the second is still prone to certain types of bugs.